src/enip/cip/message-router/index.js

```diff
@@ -14,7 +14,9 @@ const services = {
     WRITE_TAG: 0x4d,
     READ_TAG_FRAGMENTED: 0x52,
     WRITE_TAG_FRAGMENTED: 0x53,
-    READ_MODIFY_WRITE_TAG: 0x4e
+    READ_MODIFY_WRITE_TAG: 0x4e,
+    FORWARD_OPEN: 0x54,
+    FORWARD_CLOSE: 0x4E
```

Comment:
Is this the same as the `READ_MODIFY_WRITE_TAG` service? If so, that's interesting lol.

Reply:
Different object, but this is actually not good. I missed that completely. When an incoming packet gets to the _handleSendRRDataReceived() - function, how do we distinguish between the two services?

Reply:
That is tricky indeed.  Is there anything else that could be used to differentiate the packets?  Maybe packet length.  It would seem to me without digging into it that the `READ_MODIFY_WRITE TAG` response may be longer? What are your thoughts?

Reply:
Does it matter that they're the same? Can't it just emit both events? Wouldn't the controller only be listening for one at a time anyway (or if not, make it so)? The switch statement needs to be modified, but otherwise...?

Reply:
I think it will be listening to all, though you could de-register the Read_Mod listener when sending the forward close request as a potential option

Reply:
Doesn't it only subscribes on write and then removes all listeners afterwards?

Reply:
@jhenson29 Do I understand correctly that you suggest making fall-through cases for Services that have the same code?

In the current implementation (without fall-through), `READ_MODIFY_WRITE_TAG` will just never trigger, since `FORWARD_OPEN` will always be emitted to a 0x4E response-event.
What happens to the other emit without an active listener? ~~Does it get pushed to a queue?~~ -> it's synchronous. In that case it might work by just making this a fall-through case.

I'd rather implement something more elaborate. As for what a solution might be, I'm still thinking about it. At first, I thought we could just parse the IOI-Path and then decide based on that (0x06 = CM --> Must be a forward_open response), but it's not in the response-packet.

Reply:
pass through or maybe even a different way to make the decision. I didn’t code anything up yet. But as far as I know, that switch is the only (or at least primary) issue. 

I’ll look at it more during my lunch today. I have a thought on how to generalize the problem so we don’t have to worry about it. But I’m on my phone now and I don’t code on my phone.

Reply:
I didn't have time to do much testing at lunch, but in general, instead of emitting "Read Tag" for example, just emit the CIP code.
```javascript
//this.emit("Read Tag", error, data);
this.emit(service - 0x80, error, data);
```
That's almost every case of the switch statement. Special case for MSP.

Then, listen for the same.
```javascript
//this.on("Read Tag", (err, data) => {
this.on(READ_TAG, (err, data) => {
```
Then queue all requests for the same service so you don't have more than one out at the same time.

Reply:
Actually, now that I think about it, if that’s every case except MSP, then MSP code should probably be moved to the listener. I’ll have to look at it harder later. I’m on my phone again.

Reply:
I reviewed a little more. Push MSP code to another method called from the listener and just emit the CIP code w/ error and data from the existing handle method. Then, add the queues. There are queues now for 'read', 'write' and 'group' in the 'workers' property. I would guess the 'workers' property should just be a map of the CIP codes for the queues, but I didn't test.

The 'remove all listeners' calls also still need to be moved to inside the listener handlers also.

That's my 2 cents.

Reply:
Sounds like a good idea, it would facilitate synchronizing between sendRRData and sendUnitData handler. But the fix should be handled in its own issue, correct?
I'll suggest adjusting the connected-messages patch so that both 0x4E-Services emit both events, this should make things work until the handlers are adjusted, since the problem has nothing to do with the connected-patch itself.

Reply:
I would agree. Quick patch now. Another PR to correct the general issue of overlapping sevice codes.